egs/wsj/s5/utils/data/perturb_data_dir_volume.sh

```diff
@@ -25,29 +29,86 @@ if [ ! -f $data/wav.scp ]; then
+      print 'true'
+      sys.exit(0)
+  elif re.search(':[0-9]+$', line.strip()) is not None:
+    continue
+  else:
+    if ' '.join(parts[1:3]) == 'sox --vol':
+      print 'true'
+      sys.exit(0)
+print 'false'
+"` || exit 1
+
+if $volume_perturb_done; then
   echo "$0: It looks like the data was already volume perturbed.  Not doing anything."
   exit 0
 fi
 
-cat $data/wav.scp | python -c "
+if [ -z "$reco2vol" ]; then
+  cat $data/wav.scp | python -c "
 import sys, os, subprocess, re, random
 random.seed(0)
-scale_low = 1.0/8
-scale_high = 2.0
+scale_low = $scale_low
+scale_high = $scale_high
+volume_writer = open('$data/reco2vol', 'w')
 for line in sys.stdin.readlines():
   if len(line.strip()) == 0:
     continue
   # Handle three cases of rxfilenames appropriately; 'input piped command', 'file offset' and 'filename'
+  vol = random.uniform(scale_low, scale_high)
+
+  parts = line.strip().split()
   if line.strip()[-1] == '|':
-    print '{0} sox --vol {1} -t wav - -t wav - |'.format(line.strip(), random.uniform(scale_low, scale_high))
+    print '{0} sox --vol {1} -t wav - -t wav - |'.format(line.strip(), vol)
   elif re.search(':[0-9]+$', line.strip()) is not None:
-    parts = line.split()
-    print '{id} wav-copy {wav} - | sox --vol {vol} -t wav - -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = random.uniform(scale_low, scale_high))
+    print '{id} wav-copy {wav} - | sox --vol {vol} -t wav - -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = vol)
   else:
-    parts = line.split()
-    print '{id} sox --vol {vol} -t wav {wav} -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = random.uniform(scale_low, scale_high))
+    print '{id} sox --vol {vol} -t wav {wav} -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = vol)
+  volume_writer.write('{id} {vol}\n'.format(id = parts[0], vol = vol))
 "  > $data/wav.scp_scaled || exit 1;
+else
+  cat $data/wav.scp | python -c "
+import sys, os, subprocess, re
+volumes = {}
+for line in open('$reco2vol'):
+  if len(line.strip()) == 0:
+    continue
+  parts = line.strip().split()
+  volumes[parts[0]] = float(parts[1])
+
+for line in sys.stdin.readlines():
+  if len(line.strip()) == 0:
+    continue
+  # Handle three cases of rxfilenames appropriately; 'input piped command', 'file offset' and 'filename'
+  
+  parts = line.strip().split()
+  id = parts[0]
+
+  if id not in volumes:
+    raise Exception('Could not find volume for id {id}'.format(id = id))
+
+  vol = volumes[id]
+
+  if line.strip()[-1] == '|':
+    print '{0} sox --vol {1} -t wav - -t wav - |'.format(line.strip(), vol)
+  elif re.search(':[0-9]+$', line.strip()) is not None:
+    print '{id} wav-copy {wav} - | sox --vol {vol} -t wav - -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = vol)
+  else:
+    print '{id} sox --vol {vol} -t wav {wav} -t wav - |'.format(id = parts[0], wav=' '.join(parts[1:]), vol = vol)
```

Comment:
Previously it assumes that if there is sox --vol, then volume perturbation has been done. This is a bug, when reverberate_data_dir etc. is run. It has to check sox --vol is at the end of the rspecifier.